pkg/raft/term_cache_test.go

```diff
@@ -0,0 +1,396 @@
+			initialEntries := createEntriesWithFewFlips(2, 1, 1)
+			initialSlice := LogSlice{prev: entryID{term: 5, index: 10}, entries: initialEntries}
+			entries := createEntriesWithFewFlips(size, 3, 3)
+			appendSlice := LogSlice{prev: entryID{term: 5 + 1, index: 10 + 1}, entries: entries}
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				cache := NewTermCache(100, entryID{1, 1})
+				cache.truncateAndAppend(initialSlice)
+				b.StartTimer()
+
+				// This is the operation we're benchmarking
+				cache.truncateAndAppend(appendSlice)
+			}
+		})
+	}
+}
+
+func (tc *termCache) checkInvariants(t *testing.T) {
+	// check termCache is non-empty
+	require.NotEqual(t, 0, len(tc.cache), "termCache should never be empty")
+
+	// check termCache has increasing entries and terms
+	for i := 1; i < len(tc.cache); i++ {
+		if tc.cache[i].index <= tc.cache[i-1].index {
+			t.Errorf("termCache index is not strictly increasing: %d <= %d", tc.cache[i].index, tc.cache[i-1].index)
+		}
+		if tc.cache[i].term < tc.cache[i-1].term {
+			t.Errorf("termCache term is not monotonically increasing: %d < %d", tc.cache[i].term, tc.cache[i-1].term)
+		}
+	}
+}
+
+// generateRandomLogSlice creates a LogSlice with randomly generated entries for testing term cache behavior.
+// It simulates three scenarios based on random chance:
+//  1. Regular append: starts after endID
+//  2. Partial overwrite: starts between startID and endID
+//  3. Complete overwrite: starts before startID
+//
+// Parameters:
+//   - startID, endID: Define the range boundaries in the current term cache
+//   - averageLength: Controls entry count, generating between 1 and 2*averageLength entries
+//
+// The function ensures all generated entries maintain required term ordering invariants.
+func generateRandomLogSlice(startID, endID entryID, averageLength int) (lg LogSlice) {
+	if startID.index > endID.index || startID.term > endID.term {
+		panic("startID must be less than endID")
+	}
+	if startID.index == 0 || startID.term == 0 {
+		panic("startID must be higher than zero")
+	}
+	// Determine append type by random chance
+	appendType := rand.Intn(3) // 0: regular, 1: partial overwrite, 2: complete overwrite
+
+	var prevID entryID
+	switch appendType {
+	case 0: // Regular append (prev == end)
+		prevID = endID
+	case 1: // Partial overwrite (prev between start and end)
+		if startID.index == endID.index {
+			prevID = endID
+		}
+		if startID.index < endID.index {
+			randomIdx := rand.Intn(int(endID.index - startID.index))
+			prevID = entryID{term: endID.term, index: startID.index + uint64(randomIdx)}
+		}
+	case 2: // Complete overwrite (prev < start)
+		prevID = entryID{term: startID.term - 1, index: startID.index - 1}
+	}
+
+	entryCount := 1
+	if averageLength > 1 {
+		entryCount = rand.Intn(averageLength*2) + 1
+	}
+
+	// Generate terms with possible term flips
+	terms := make([]uint64, entryCount)
+	currentTerm := prevID.term + 1 // Start with term greater than prev
```

Comment:
It's typical that the term does not change, don't need to start with a higher one.